air-quality-ui/system_tests/ui/summary_page.spec.ts

```diff
@@ -1,75 +1,807 @@
+        overall_aqi_level: { mean: 2 },
+        no2: { mean: { aqi_level: 1, value: 1.2220194497781245 } },
+        o3: { mean: { aqi_level: 3, value: 110.29793453644987 } },
+        pm2_5: { mean: { aqi_level: 2, value: 15.764618078867594 } },
+        pm10: { mean: { aqi_level: 2, value: 31.71802172436572 } },
+        so2: { mean: { aqi_level: 1, value: 1.3459434727665889 } },
+      }),
+    ]
+    await vairifySummaryPage.setupPageWithMockData(
+      mockedForecastResponse,
+      mockedMeasurementSummaryResponse,
+    )
+  })
+  test.describe('Table Structure', () => {
+    test('Verify that Headers exist and Innertext matches', async ({
+      vairifySummaryPage,
+    }) => {
+      await vairifySummaryPage.getColumnHeaderAndText('AQI Level', 'AQI Level')
+      await vairifySummaryPage.getColumnHeaderAndText(
+        'PM 2.5 (µg/m³)',
+        'PM 2.5 (µg/m³)',
+      )
+      await vairifySummaryPage.getColumnHeaderAndText(
+        'PM 10 (µg/m³)',
+        'PM 10 (µg/m³)',
+      )
+      await vairifySummaryPage.scrollToRightmostPosition()
+      await vairifySummaryPage.page.waitForTimeout(1000)
+
+      await vairifySummaryPage.getColumnHeaderAndText(
+        'Nitrogen Dioxide (µg/m³)',
+        'Nitrogen Dioxide (µg/m³)',
+      )
+      await vairifySummaryPage.getColumnHeaderAndText(
+        'Ozone (µg/m³)',
+        'Ozone (µg/m³)',
+      )
+      await vairifySummaryPage.getColumnHeaderAndText(
+        'Sulphur Dioxide (µg/m³)',
+        'Sulphur Dioxide (µg/m³)',
+      )
+    })
+
+    test('Verify numbers in cells have no more than 1 decimal place ', async ({
+      vairifySummaryPage,
+    }) => {
+      await vairifySummaryPage.checkCellNumberFormat()
+    })
+  })
+
+  test.describe('Table Data Validation', () => {
+    test('Verify that a city with no in-situ data still show on grid', async ({
+      vairifySummaryPage,
+    }) => {
+      const count = await vairifySummaryPage.textCellSearch('Kyiv')
+      expect(count).toEqual(1)
+    })
+
+    test('Check data is displayed correctly on grid', async ({
+      vairifySummaryPage,
+    }) => {
+      // first 6 values for Kampala, Abu Dhabi, Zurich and Kyiv respectively
+      const expectedData = [
+        ['2', '6', '-4', '16.1', '76', '19 Jun 09:00'],
+        ['4', '5', '-1', '30.3', '52.8', '19 Jun 12:00'],
+        ['2', '2', '0', '17.2', '15.8', '19 Jun 12:00'],
+        ['2', '-', '-', '7', '-', '24 Jun 09:00'],
+      ]
+
+      await vairifySummaryPage.assertGridValues(expectedData)
+    })
+
+    test('Verify that Diff displays the delta between forcast and measured', async ({
+      vairifySummaryPage,
+    }) => {
+      const diffArray = await vairifySummaryPage.calculateForecastDifference()
+      expect(diffArray).toEqual([-4, -1, 0])
+    })
+  })
 })
 
-test.describe('Mocked API tests', () => {
-  test('Verify that Headers exist and Innertext matches', async ({
+test.describe('Table Data: Largest Deviation', () => {
```

Comment:
Unless I'm missing something (possible) we  have 5 tests that are all basically testing the same thing with slightly different numbers. We should look into parameterising these if we can, or possibly reducing the number. 
How much benefit is there in having all 5?

Reply:
I tried to parameterize these, but unlike in pytest I couldn't get it to work with passing in objects rather than just strings as I did in the BVA tests at the top of the file. It might be possible, but I had a look with a dev to see if I was missing something at the time. I was trying to check that this worked whether the forecast was larger, or the measurement, and the difference in the test with aqi 4 vs 3, and aqi 6 vs 3 is that the latter has other pollutants with a diff and so I wanted to check that it picked the highest. Maybe I should just remove the 4 vs 3 ones